src/Rector/AbstractRector.php

```diff
@@ -73,89 +60,15 @@ public function setBetterNodeFinder(BetterNodeFinder $betterNodeFinder): void
      */
     public function afterTraverse(array $nodes): array
     {
-        return $this->prependExpressionNodes($nodes);
-    }
-
-    protected function prependNodeAfterNode(Expr $nodeToPrepend, Node $positionNode): void
-    {
-        $positionExpressionNode = $this->betterNodeFinder->findFirstAncestorInstanceOf(
-            $positionNode,
-            Expression::class
-        );
-
-        $expressionToPrepend = $this->wrapToExpression($nodeToPrepend);
-
-        if (isset($this->expressionsToPrependAfter[$positionExpressionNode])) {
-            $this->expressionsToPrependAfter[$positionExpressionNode] = array_merge(
-                $this->expressionsToPrependAfter[$positionExpressionNode],
-                [$expressionToPrepend]
-            );
-        } else {
-            $this->expressionsToPrependAfter[$positionExpressionNode] = [$expressionToPrepend];
-        }
-    }
-
-    protected function prependNodeBeforeNode(Expr $nodeToPrepend, Node $positionNode): void
-    {
-        $positionExpressionNode = $this->betterNodeFinder->findFirstAncestorInstanceOf(
-            $positionNode,
-            Expression::class
-        );
-
-        $expressionToPrepend = $this->wrapToExpression($nodeToPrepend);
-
-        if (isset($this->expressionsToPrependBefore[$positionExpressionNode])) {
-            $this->expressionsToPrependBefore[$positionExpressionNode] = array_merge(
-                $this->expressionsToPrependBefore[$positionExpressionNode],
-                [$expressionToPrepend]
-            );
-        } else {
-            $this->expressionsToPrependBefore[$positionExpressionNode] = [$expressionToPrepend];
-        }
-    }
-
-    /**
-     * Adds new nodes before or after particular Expression nodes.
-     *
-     * @param Node[] $nodes
-     * @return Node[] array
-     */
-    private function prependExpressionNodes(array $nodes): array
-    {
-        foreach ($nodes as $i => $node) {
-            if ($node instanceof Expression) {
-                $nodes = $this->prependNodesAfterAndBeforeExpression($nodes, $node, $i);
-            } elseif (isset($node->stmts)) {
-                $node->stmts = $this->prependExpressionNodes($node->stmts);
-            }
-        }
-
-        return $nodes;
-    }
-
-    /**
-     * @param Node[] $nodes
-     * @return Node[]
-     */
-    private function prependNodesAfterAndBeforeExpression(array $nodes, Node $node, int $i): array
-    {
-        if (isset($this->expressionsToPrependBefore[$node])) {
-            array_splice($nodes, $i, 0, $this->expressionsToPrependBefore[$node]);
-
-            unset($this->expressionsToPrependBefore[$node]);
-        }
+        // prepend new expressions
+        $nodeTraverser = new NodeTraverser();
+        $nodeTraverser->addVisitor($this->expressionPrependingNodeVisitor);
```

Comment:
Maybe move to `expressionPrepender` service?

Btw, note to myself: always use `NodeTraverser`, it checks all cases much better than anything custom.

Also `array_splice()` usually mean there should be NodeTraverser with `leaveNode(): array` returning the old + added node